File: Java Data/T 01 Patterns/Pattern_09_Diamond_Star_Pattern.java
import java.util.Scanner;

//     *
//    ***
//   *****
//  *******
// *********
// *********
//  *******
//   *****
//    ***
//     *

public class Pattern_09_Diamond_Star_Pattern {

    /**
     * Prints the top half (a standard centered pyramid) of the diamond pattern
     */
    public static void starPyramidPattern(int n) {
        // Loops through each row for the top half
        for (int i = 1; i <= n; i++) {

            // Prints leading spaces
            for (int j = 1; j <= n - i; j++) {
                System.out.print(" ");
            }

            // Prints stars (1, 3, 5, etc.).
            for (int j = 1; j <= (2 * i - 1); j++) {
                System.out.print("*");
            }

            // Prints trailing spaces
            for (int j = 1; j <= n - i; j++) {
                System.out.print(" ");
            }
            // Prints a newline after the row is done
            System.out.println();
        }
    }

    /**
     * Prints the bottom half (an inverted centered pyramid) of the diamond pattern
     */
    public static void invertedStarPyramidPattern(int n) {
        // Loops through each row for the bottom half
        for (int i = 1; i <= n; i++) {

            // Prints leading spaces
            for (int j = 1; j <= i - 1; j++) {
                System.out.print(" ");
            }

            // Prints stars (total 2*n - (2*i - 1))
            for (int j = 1; j <= 2 * n - (2 * i - 1); j++) {
                System.out.print("*");
            }
            // Prints trailing spaces
            for (int j = 1; j <= i - 1; j++) {
                System.out.print(" ");
            }

            // Prints a newline after the row is done
            System.out.println();
        }
    }

    /**
     * Combines the top and bottom halves to print a complete diamond pattern
     */
    public static void diamondStarPyramidPattern(int n) {
        starPyramidPattern(n);
        invertedStarPyramidPattern(n);
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of rows for the Diamond star pyramid pattern: ");
        int n = scanner.nextInt();

        diamondStarPyramidPattern(n);

        scanner.close();
    }
}

File: 00-Patterns-Python/Pattern_09_Diamond_Star_Pattern.py
#     *
#    ***
#   *****
#  *******
# *********
# *********
#  *******
#   *****
#    ***
#     *

def star_pyramid_pattern(n):
    for i in range(1, n + 1, 1):
        for j in range(1, n - i + 1, 1):
            print(" ", end="")
        for j in range(1, i * 2, 1):
            print("*", end="")
        print()

        
def inverted_star_pyramid_pattern(n):
    for i in range(1, n + 1, 1):
        for j in range(1, i, 1):
            print(" ", end="")
        for j in range(1, 2 * (n - i + 1) , 1):
            print("*", end="")
        print()


# Combines the top and bottom halves to print a complete diamond pattern
def diamond_pattern(n):
    star_pyramid_pattern(n)
    inverted_star_pyramid_pattern(n)

num = input("Enter the number of rows for the Diamond star pyramid pattern: ")
diamond_pattern(int(num))
